File: leetcode/py/70-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = {}
        
        def helper(n):
            if n < 1:
                return 0
            elif n == 1:
                return 1
            elif n == 2:
                return 2
            elif n in memo:
                return memo[n]
            else:
                memo[n] = helper(n - 1) + helper(n - 2)
                return memo[n]
        
        return helper(n)

x = Solution()
print(x.climbStairs(10))
print(x.climbStairs(15))
print(x.climbStairs(40))

File: leetcode/js/70-climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */

function helper(n, memo) {
    if (n === 0) {
        return 0;
    }
    else if (n === 1) {
        return 1;
    }
    else if (n === 2) {
        return 2;
    }
    else if (memo.has(n)) {
        return memo.get(n);
    }
    else{
        memo.set(n, helper(n - 1, memo) + helper(n - 2, memo));
        return memo.get(n);
    }
}

var climbStairs = function(n) {
    let memo = new Map()

    return helper(n, memo);
};